plugins/inputs/sflow/packetdecoder/packetdecoder.go

```diff
@@ -0,0 +1,506 @@
+		fr := types.FlowRecord{}
+		if err := read(r, &fr.FlowFormat, "FlowFormat"); err != nil { // line 1597
+			return recs, err
+		}
+		if err := read(r, &flowDataLen, "Flow data length"); err != nil {
+			return recs, err
+		}
+
+		if flowDataLen > MaxFlowDataLen {
+			return recs, fmt.Errorf("sample data length too large")
+		}
+		b := make([]byte, flowDataLen)
+		if err := read(r, &b, "Flow Data"); err != nil {
+			return recs, err
+		}
+		buf := bytes.NewBuffer(b)
+
+		switch fr.FlowFormat {
+		case types.FlowFormatTypeRawPacketHeader: // line 1938
+			fr.FlowData, err = d.decodeRawPacketHeaderFlowData(buf, samplingRate)
+		default:
+			d.debug("Unknown flow format: ", fr.FlowFormat)
+		}
+		if err != nil {
+			return recs, err
+		}
+
+		recs = append(recs, fr)
+	}
+
+	return recs, err
+}
+
+func (d *PacketDecoder) decodeRawPacketHeaderFlowData(r io.Reader, samplingRate uint32) (h types.RawPacketHeaderFlowData, err error) {
+	if err := read(r, &h.HeaderProtocol, "HeaderProtocol"); err != nil { // line 1940
+		return h, err
+	}
+	if err := read(r, &h.FrameLength, "FrameLength"); err != nil { // line 1942
+		return h, err
+	}
+	h.Bytes = h.FrameLength * samplingRate
+
+	if err := read(r, &h.StrippedOctets, "StrippedOctets"); err != nil { // line 1967
+		return h, err
+	}
+	if err := read(r, &h.HeaderLength, "HeaderLength"); err != nil {
+		return h, err
+	}
+
+	if h.HeaderLength > MaxPacketHeaderLen {
+		return h, fmt.Errorf("packet header length too large")
+	}
+
+	b := make([]byte, h.HeaderLength)
+	read(r, &b, "header data")
+	buf := bytes.NewBuffer(b)
+
+	switch h.HeaderProtocol {
+	case types.HeaderProtocolTypeEthernetISO88023:
+		h.Header, err = d.decodeEthHeader(buf)
+	default:
+		d.debug("Unknown header protocol type: ", h.HeaderProtocol)
+	}
+
+	return h, err
+}
+
+// ethHeader answers a decode Directive that will decode an ethernet frame header
+// according to https://en.wikipedia.org/wiki/Ethernet_frame
+func (d *PacketDecoder) decodeEthHeader(r io.Reader) (h types.EthHeader, err error) {
+	// we may have to read out StrippedOctets bytes and throw them away first?
+	if err := read(r, &h.DestinationMAC, "DestinationMAC"); err != nil {
+		return h, err
+	}
+	if err := read(r, &h.SourceMAC, "SourceMAC"); err != nil {
+		return h, err
+	}
+	var tagOrEType uint16
+	if err := read(r, &tagOrEType, "tagOrEtype"); err != nil {
+		return h, err
+	}
+	switch tagOrEType {
+	case 0x8100: // could be?
+		var discard uint16
+		if err := read(r, &discard, "unknown"); err != nil {
+			return h, err
+		}
+		if err := read(r, &h.EtherTypeCode, "EtherTypeCode"); err != nil {
+			return h, err
+		}
+	default:
+		h.EtherTypeCode = tagOrEType
+	}
+	h.EtherType = types.ETypeMap[h.EtherTypeCode]
+	switch h.EtherType {
+	case "IPv4":
+		h.IPHeader, err = d.decodeIPv4Header(r)
+	case "IPv6":
+		h.IPHeader, err = d.decodeIPv6Header(r)
+	default:
+	}
+	if err != nil {
+		return h, err
+	}
+	// missing reading crc check bytes?
```

Comment:
Let's figure this out and improve the comment.

Reply:
Pretty sure we're missing more than a couple things and we just don't care about it. The MinReader behavior covers this.